File: solutions/csharp/camicia/1/Camicia.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class Camicia
{
    public enum GameStatus
    {
        Finished,
        Loop
    }

    public record GameResult(GameStatus Status, int Tricks, int Cards);

    public static GameResult SimulateGame(string[] playerA, string[] playerB)
    {
        var deckA = new Queue<string>(playerA);
        var deckB = new Queue<string>(playerB);
        var pile = new List<string>();

        int cardsPlayed = 0;
        int tricks = 0;

        int player = 0;
        int penalty = 0;
        int lastFacePlayer = -1;

        var seen = new HashSet<string>();

        while (true)
        {
            // LOOP CHECK (solo al inicio de ronda)
            if (penalty == 0 && pile.Count == 0)
            {
                string state = Serialize(deckA, deckB, player, penalty);

                if (!seen.Add(state))
                    return new GameResult(GameStatus.Loop, tricks, cardsPlayed);
            }

            var deck = player == 0 ? deckA : deckB;

            if (deck.Count == 0)
            {
                var winner = player == 0 ? deckB : deckA;
                Collect(winner, pile);
                tricks++;

                if (deckA.Count == 0 || deckB.Count == 0)
                    return new GameResult(GameStatus.Finished, tricks, cardsPlayed);

                player = lastFacePlayer;
                continue;
            }

            var card = deck.Dequeue();
            pile.Add(card);
            cardsPlayed++;

            int p = Penalty(card);

            if (p > 0)
            {
                penalty = p;
                lastFacePlayer = player;
                player = 1 - player;
                continue;
            }

            if (penalty > 0)
            {
                penalty--;

                if (penalty == 0)
                {
                    var winner = lastFacePlayer == 0 ? deckA : deckB;
                    Collect(winner, pile);
                    tricks++;

                    if (deckA.Count == 0 || deckB.Count == 0)
                        return new GameResult(GameStatus.Finished, tricks, cardsPlayed);

                    player = lastFacePlayer;
                }
            }
            else
            {
                player = 1 - player;
            }
        }
    }

    static int Penalty(string card)
    {
        return card switch
        {
            "J" => 1,
            "Q" => 2,
            "K" => 3,
            "A" => 4,
            _ => 0
        };
    }

    static void Collect(Queue<string> deck, List<string> pile)
    {
        foreach (var c in pile)
            deck.Enqueue(c);

        pile.Clear();
    }

    // NUEVO SERIALIZE (corrige los loops falsos)
    static string Serialize(Queue<string> a, Queue<string> b, int player, int penalty)
    {
        string sa = string.Join(",", a.Select(Normalize));
        string sb = string.Join(",", b.Select(Normalize));

        return $"{player}:{penalty}:{sa}|{sb}";
    }

    static string Normalize(string card)
    {
        return card switch
        {
            "J" => "J",
            "Q" => "Q",
            "K" => "K",
            "A" => "A",
            _ => "N"
        };
    }
}

File: solutions/java/camicia/1/src/main/java/Camicia.java
import java.util.*;
public class Camicia {
    private static final String FINISHED = "finished";
    private static final String LOOP = "loop";
    public static CamiciaResult simulateGame(List<String> pA, List<String> pB) {
        List<String> playerA = new ArrayList<>(pA);
        List<String> playerB = new ArrayList<>(pB);
        List<String> stack = new ArrayList<>();
        int gameTricks = 0;
        int totalCardsPlayed = 0;
        int totalCards = playerA.size() + playerB.size();
        List<String> penalty = List.of("0", "J", "Q", "K", "A");
        String turn = "A";
        Set<String> visitedStates = new HashSet<>();
        while (!playerA.isEmpty() || !playerB.isEmpty()) {
            String state = getSignature(playerA, playerB, turn);
            if (visitedStates.contains(state)) {
                return new CamiciaResult(LOOP, totalCardsPlayed, gameTricks);
            }
            visitedStates.add(state);
            String card;
            if (turn.equals("A")) {
                card = playerA.isEmpty() ? null : playerA.removeFirst();
            } else {
                card = playerB.isEmpty() ? null : playerB.removeFirst();
            }
            if (card == null) {
                List<String> winner = turn.equals("A") ? playerB : playerA;
                winner.addAll(stack);
                stack.clear();
                gameTricks++;
                break;
            }
            totalCardsPlayed++;
            stack.add(card);
            int penaltyIndex = penalty.indexOf(card);
            if (penaltyIndex > 0) {
                boolean isPenaltyActive = true;
                String target = turn.equals("A") ? "B" : "A";
                int count = penaltyIndex;
                while (isPenaltyActive) {
                    boolean responded = false;
                    for (int i = 0; i < count; i++) {
                        String pCard;
                        if (target.equals("A")) {
                            pCard = playerA.isEmpty() ? null : playerA.removeFirst();
                        } else {
                            pCard = playerB.isEmpty() ? null : playerB.removeFirst();
                        }
                        if (pCard == null) {
                            isPenaltyActive = false;
                            break;
                        }
                        totalCardsPlayed++;
                        stack.add(pCard);
                        int newPenalty = penalty.indexOf(pCard);
                        if (newPenalty > 0) {
                            count = newPenalty;
                            target = target.equals("A") ? "B" : "A";
                            responded = true;
                            break;
                        }
                    }
                    if (!responded) {
                        isPenaltyActive = false;
                        String winnerName = target.equals("A") ? "B" : "A";
                        if (winnerName.equals("A")) {
                            playerA.addAll(stack);
                        } else {
                            playerB.addAll(stack);
                        }
                        stack.clear();
                        gameTricks++;
                        turn = winnerName;
                    }
                }
            } else {
                turn = turn.equals("A") ? "B" : "A";
            }
            if (playerA.size() == totalCards || playerB.size() == totalCards) {
                break;
            }
        }
        if (!stack.isEmpty()) {
            gameTricks++;
            List<String> winner = playerA.isEmpty() ? playerB : playerA;
            winner.addAll(stack);
        }
        boolean isLoop =
                playerA.size() != totalCards &&
                        playerB.size() != totalCards;
        return new CamiciaResult(
                isLoop ? LOOP : FINISHED,
                totalCardsPlayed,
                gameTricks
        );
    }
    private static String getSignature(
            List<String> deckA,
            List<String> deckB,
            String currentTurn
    ) {
        return norm(deckA) + "|" + norm(deckB) + "|" + currentTurn;
    }
    private static String norm(List<String> deck) {
        StringBuilder sb = new StringBuilder();
        for (String card : deck) {
            if (card.matches("^\\d+$")) {
                sb.append("N");
            } else {
                sb.append(card);
            }
        }
        return sb.toString();
    }
}
